nyxtools/vector.py

```diff
@@ -0,0 +1,253 @@
+    # Only needed if hold==True
+    proceed = Cpt(EpicsSignal, "Cmd:Proceed-Cmd")
+
+    # Abort the current vector motion
+    abort = Cpt(EpicsSignal, "Cmd:Abort-Cmd")
+
+    # Set all vector motors start and end position to their current RBV values
+    sync = Cpt(EpicsSignal, "Cmd:Sync-Cmd")
+
+    def run(
+        self,
+        o: Tuple[float,float], x: Tuple[float,float], y: Tuple[float, float], z: Tuple[float,float],
+        exposure: float, num_samples: float, buffer_time: float, shutter_lag_time: float, shutter_time: float):
+
+        # Configure motion
+        yield from bps.abs_set(self.sync, 1, wait=True)
+
+        yield from bps.mv(
+            self.calc_only, True, # Check for errors first
+            self.expose, True,
+            self.hold, False,
+
+            self.exposure, exposure,
+            self.num_samples, num_samples,
+            self.buffer_time, buffer_time,
+            self.shutter_lag_time, shutter_lag_time,
+            self.shutter_time, shutter_time,
+
+            self.o.start, o[0],
+            self.o.end, o[1],
+
+            self.x.start, x[0],
+            self.x.end, x[1],
+
+            self.y.start, y[0],
+            self.y.end, y[1],
+
+            self.z.start, z[0],
+            self.z.end, z[1],
+
+            group='vec_config'
+        )
+
+        yield from bps.wait('vec_config')
+
+        # Start "motion"
+        yield from bps.abs_set(self.go, 1, wait=False)
+
+        # There's no way to know it is done, so wait a little, it should be very fast
+        yield from bps.sleep(1.0)
+
+        # Check for errors
+        error = yield from bps.rd(self.error, as_string=True, use_monitor=False)
+
+        if error != "None":
+            raise Exception(f"Failed to run vector. Error: {error}")
+
+        # Estimate total motion time (in ms)
+        time_to_speed = yield from bps.rd(self.max_time_to_speed)
+        buffer_time = yield from bps.rd(self.buffer_time)
+        shutter_time = yield from bps.rd(self.shutter_time)
+        daq_duration = yield from bps.rd(self.data_acq_duration)
+
+        estimated_total_time_ms = 2*time_to_speed + buffer_time + 2*shutter_time + daq_duration
+        timeout = 5*estimated_total_time_ms/1000.0
+
+        # Start actual motion
+        yield from bps.abs_set(self.calc_only, False, wait=True)
+        yield from bps.abs_set(self.go, 1, wait=False)
+
+        # Wait until it is done
+
+        # NOTE: ideally we should catch self.running going from 0->1 and then from 1->0
+        # but it is not guaranteed that we can observe these two transitions.
+        # instead, we wait a little bit after the motion is started (hopefully past 0->1)
+        # and then wait until either we see 1->0 or a timeout expires
+
+        yield from bps.sleep(0.2)
+
+        t = time.time()
+
+        while True:
+            running = yield from bps.rd(self.running)
+            elapsed = time.time() - t
+
+            if not running or elapsed > timeout:
```

Comment:
If elapsed > timeout should we throw a warning/error? I'm not sure how bad it would be to try to run it again while it's already running, or how likely (or frequently) this is to happen.

Reply:
This `run` plan is intended more as an example of how to use the vector object in a plan. Keep in mind that the Vector is rarely run by itself. The idea is to do a scan as follows:

1. Configure detector + Zebra + Vector ("Hold"=true)
2. Start the Vector (It will backup the motors and wait. This is done to prevent issues with the Zebra capturing spurious triggers during the backing up)
3. Arm the detector + Zebra
4. Resume the vector, which will perform the actual movement

By this point you have two ways to know the vector is done: the detector is disarmed and Zebra is done.

I expect LSDC to have a more complicated plan doing the steps above involving all 3 ophyd objects.

Reply:
I am happy with the way the current run plan works for the purposes of this pull request. We may restructure this later, but it has worked well enough for us to test the code and IOC.